### PR TITLE
New debug framework

### PR DESCRIPTION
**Pull request by pradeepns85**
This has 2 commits . kindly review
1.  os side changes for debug framework
2. application changes for using debug framework

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/debug/Kconfig [hunk 1]
```diff
@@ -0,0 +1,7 @@
+menu "System Debug Application"
```
Comment (sunghan-chang) [inline apps/debug/Kconfig]: Please add belows
```
#
# For a description of the syntax of this configuration file,
# see kconfig-language at https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt
#
```

File: os/arch/arm/src/armv7-r/arm_blocktask.c [hunk 2]
```diff
@@ -67,6 +67,10 @@
 #include "group/group.h"
 #include "up_internal.h"
 
+#if defined(CONFIG_DEBUG_SYSTEM) && defined(CONFIG_TASK_SCHED_HISTORY)
```
Comment (sunghan-chang) [inline os/arch/arm/src/armv7-r/arm_blocktask.c]: ```defined(CONFIG_DEBUG_SYSTEM)``` is not needed. ```#if defined(CONFIG_TASK_SCHED_HISTORY)```  is enough. Because CONFIG_TASK_SCHED_HISTORY will be enabled only under enabled CONFIG_DEBUG_SYSTEM.
All of these kinds checking are same.

File: os/include/tinyara/debug/sysdbg.h [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+#include <tinyara/config.h>
+#include <tinyara/sched.h>
+#include <tinyara/clock.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#ifdef CONFIG_ARCH_ARM
```
Comment (sunghan-chang) [inline os/include/tinyara/debug/sysdbg.h]: There is same code on mm.h. We need to move (or make) this to header of arch.

File: os/arch/arm/src/armv7-r/arm_blocktask.c (hunk 2 above)
Comment (pradeepns85) [reply os/arch/arm/src/armv7-r/arm_blocktask.c]: done

File: os/include/tinyara/debug/sysdbg.h (hunk 3 above)
Comment (pradeepns85) [reply os/include/tinyara/debug/sysdbg.h]: taken care

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by sunghan-chang** [comment]
How about move apps/debug to apps/system/debug?

**Comment by sunghan-chang** [comment]
And debug is too big topic. How about simulator or monitor or something else?

File: apps/debug/Makefile [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+endif
+
+ifeq ($(WINTOOL),y)
+  INSTALL_DIR = "${shell cygpath -w $(BIN_DIR)}"
+else
+  INSTALL_DIR = $(BIN_DIR)
+endif
+
+CONFIG_EXAMPLES_DEBUG_PROGNAME ?= debug$(EXEEXT)
+PROGNAME = $(CONFIG_EXAMPLES_BENCHMARK_PROGNAME)
```
Comment (sunghan-chang) [inline apps/debug/Makefile]: check program config name. mismatched with above line

File: apps/debug/sysdbgapp_init.c [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+ *   int - zero indicates successful execution
+ *
+ * Assumptions:
+ *   None
+ *
+ ****************************************************************************/
+
+int sysdbgapp_main(int argc, char **args)
+{
+	int i, nbytes;
```
Comment (sunghan-chang) [inline apps/debug/sysdbgapp_init.c]: Let's define at different line for nbytes.

File: apps/debug/sysdbgapp_init.c [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+	char arg_str[128];
+	if (argc < 2) {
+		show_usage(args[0]);
+	} else if (strcmp("read", args[1]) == 0) {
```
Comment (sunghan-chang) [inline apps/debug/sysdbgapp_init.c]: how about use strncmp to prevent security issue on code analyze tool?

File: apps/debug/sysdbgapp_init.c [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+		sysdbg_dev_read();
+	} else {
+		/* Open system debug device created in kernel/debug/sysdbg.c */
+		if (sysdbg_dev_open() < 0) {
+			return 0;
```
Comment (sunghan-chang) [inline apps/debug/sysdbgapp_init.c]: return 0?

File: apps/debug/sysdbgapp_init.c [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+		}
+
+		if (strcmp("max_task_count", args[1]) == 0 || strcmp("max_irq_count", args[1]) == 0 || strcmp("max_sem_count", args[1]) == 0) {
+			i = 1;
+			memset((void *)arg_str, 0, strlen(arg_str));
+			while (argc > 1) {
+				if (i == 1) {
+					sprintf(arg_str, "%s", args[i]);
```
Comment (sunghan-chang) [inline apps/debug/sysdbgapp_init.c]: how about snprintf?

File: apps/debug/sysdbgapp_init.c [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+		show_usage(args[0]);
+	} else if (strcmp("read", args[1]) == 0) {
+		sysdbg_dev_read();
+	} else {
+		/* Open system debug device created in kernel/debug/sysdbg.c */
+		if (sysdbg_dev_open() < 0) {
+			return 0;
+		}
+
+		if (strcmp("max_task_count", args[1]) == 0 || strcmp("max_irq_count", args[1]) == 0 || strcmp("max_sem_count", args[1]) == 0) {
```
Comment (sunghan-chang) [inline apps/debug/sysdbgapp_init.c]: strncmp?

File: apps/debug/sysdbgapp_init.c [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+				} else {
+					sprintf(arg_str + strlen(arg_str), " %s", args[i]);
+				}
+				argc--;
+				i++;
+			}
+			nbytes = write(g_sysdbg_handle, arg_str, strlen(arg_str));
+		} else {
+			nbytes = write(g_sysdbg_handle, args[1], strlen(args[1]));
+
```
Comment (sunghan-chang) [inline apps/debug/sysdbgapp_init.c]: remove this empty line

File: apps/include/debug/sysdbgapp_init.h [hunk 11]
```diff
@@ -0,0 +1,44 @@
+/****************************************************************************
+ *
+ * Copyright (c) 2016 Samsung Electronics Co., Ltd All Rights Reserved
```
Comment (sunghan-chang) [inline apps/include/debug/sysdbgapp_init.h]: 2017

File: apps/include/debug/sysdbgapp_init.h [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ ****************************************************************************
+ *
+ * WARNING: All information contained herein is, and remains the property of
```
Comment (sunghan-chang) [inline apps/include/debug/sysdbgapp_init.h]: Here is proprietary license. Is it right?

File: os/kernel/debug/sysdbg.c [hunk 13]
```diff
@@ -0,0 +1,941 @@
...
+#include <tinyara/config.h>
+#include <tinyara/arch.h>
+#include <tinyara/kmalloc.h>
+#include <tinyara/debug/sysdbg.h>
+#include "sched/sched.h"
+
+/* Kindly update the debug_version as and when changes done in struct
+ * sysdbg_struct as there will be dependency on T32 scripts and
+ * python script to handle the structure */
+volatile int g_debug_version = 1;
```
Comment (sunghan-chang) [inline os/kernel/debug/sysdbg.c]: What's the usages of this? Now, this is not called by anyplaces.

File: apps/debug/sysdbgapp_init.c [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+					sprintf(arg_str, "%s", args[i]);
+				} else {
+					sprintf(arg_str + strlen(arg_str), " %s", args[i]);
+				}
+				argc--;
+				i++;
+			}
+			nbytes = write(g_sysdbg_handle, arg_str, strlen(arg_str));
+		} else {
+			nbytes = write(g_sysdbg_handle, args[1], strlen(args[1]));
```
Comment (sunghan-chang) [inline apps/debug/sysdbgapp_init.c]: You split locations of parsing argurment. How about do it in same logic?

File: os/Kconfig.debug [hunk 15]
```diff
@@ -529,6 +529,45 @@ config DEBUG_TTRACE
 	---help---
 		Enable T-trace debug SYSLOG output (disabled by default)
 
+comment "System Debug Options"
+
+config DEBUG_SYSTEM
+	bool "System debug"
+	default n
+	---help---
```
Comment (sunghan-chang) [inline os/Kconfig.debug]: You should make a dependancy with DEBUG. Because you use "dbg" in kernel/debug.

File: os/kernel/Kconfig [hunk 16]
```diff
@@ -1040,4 +1040,3 @@ config PTHREAD_STACK_DEFAULT
 	---help---
 		Default pthread stack size
 endmenu # Stack size information
-
```
Comment (sunghan-chang) [inline os/kernel/Kconfig]: let's revert this change.

**Comment by pradeepns85** [comment]
@sunghan-chang , Regarding comment "How about move apps/debug to apps/system/debug?"   Since debug folder will be moved inside system folder, scope of  debug name would be limited to system debug. 
As you know this application can be enhanced to debug any of the modules inside os for ex.  log list of tasks/threads/functions accessing a specific memory region , or monitoring the mutex, or monitoring any other system resources acesss

As you said, we can move the debug folder under  apps/system/debug  and can keep the same folder name. what's your opinion

File: apps/debug/Makefile [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+#
+###########################################################################
+
+-include $(TOPDIR)/.config
+-include $(TOPDIR)/Make.defs
+include $(APPDIR)/Make.defs
+
+# System Debugging Application
+
+APPNAME = sysdbgapp
```
[the remark this answers is not in the document]
Comment (pradeepns85) [reply apps/debug/Makefile]: done

File: apps/debug/Makefile [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+
+# Test Case Example
+
+ASRCS =
+CSRCS =
+ifeq ($(CONFIG_DEBUG_SYSTEM_APP),y)
```
[the remark this answers is not in the document]
Comment (pradeepns85) [reply apps/debug/Makefile]: done

File: apps/debug/Makefile (hunk 4 above)
Comment (pradeepns85) [reply apps/debug/Makefile]: done

File: apps/debug/sysdbgapp_init.c (hunk 5 above)
Comment (pradeepns85) [reply apps/debug/sysdbgapp_init.c]: done

File: apps/debug/sysdbgapp_init.c (hunk 6 above)
Comment (pradeepns85) [reply apps/debug/sysdbgapp_init.c]: done

File: apps/debug/sysdbgapp_init.c (hunk 7 above)
Comment (pradeepns85) [reply apps/debug/sysdbgapp_init.c]: changed to -1

File: apps/debug/sysdbgapp_init.c (hunk 9 above)
Comment (pradeepns85) [reply apps/debug/sysdbgapp_init.c]: done

File: apps/debug/sysdbgapp_init.c (hunk 8 above)
Comment (pradeepns85) [reply apps/debug/sysdbgapp_init.c]: done

File: apps/debug/sysdbgapp_init.c (hunk 14 above)
Comment (pradeepns85) [reply apps/debug/sysdbgapp_init.c]: handled in single logic

File: apps/debug/sysdbgapp_init.c (hunk 10 above)
Comment (pradeepns85) [reply apps/debug/sysdbgapp_init.c]: done

File: apps/debug/sysdbgapp_init.c (hunk 14 above)
Comment (sunghan-chang) [reply apps/debug/sysdbgapp_init.c]: No.
```
+	dbg("USAGE:\n");
 +	dbg("sysdbg enable  /* Allocates memory and enables logging */\n"); <- os
 +	dbg("sysdbg disable /* Frees the memory and disables logging */\n"); <- os
 +	dbg("sysdbg read   /* Reads /dev/sysdbg to view debug information*/\n"); <- apps
 +	dbg("sysdbg max_task_count <count> /* Enter max_task_count */\n"); <- apps
 +	dbg("sysdbg max_irq_count <count> /* Enter max_irq_count */\n"); <- apps
 +	dbg("sysdbg max_sem_count <count> /* Enter max_sem_count */\n");<- apps
 +	dbg("sysdbg data_abort /* Simulate data abort */\n"); <- os
 +	dbg("sysdbg prefetch_abort /* Simulate Prefetch abort */\n"); <- os
 +	dbg("sysdbg usr_assert /* Simulate User Assert */\n"); <- os
 +	dbg("sysdbg undef_abort  /* Simulate Undefined abort */\n"); <- os
 +	dbg("sysdbg dump_stack /* Dumps the current task stack */\n"); <- os
 +	dbg("sysdbg dump_allstack /* Dumps the stack of all tasks */\n"); <- os
```

File: os/kernel/debug/sysdbg.c (hunk 13 above)
Comment (pradeepns85) [reply os/kernel/debug/sysdbg.c]: This will be used in future by python script to parse the debug information based on the version

**Comment by sunghan-chang** [comment]
@pradeepns85 I don't think system means hw fucntionality. I think that is an application provided by tizenrt (us), not user application. Vi is there as well. That's why suggest.

File: apps/debug/sysdbgapp_init.c (hunk 14 above)
Comment (pradeepns85) [reply apps/debug/sysdbgapp_init.c]: Sorry, I mis understood your comment earlier. I will handle it

File: os/Kconfig.debug (hunk 15 above)
Comment (pradeepns85) [reply os/Kconfig.debug]: DEBUG_SYSTEM is already inside DEBUG , however i added dependency with DEBUG_VERBOSE
if DEBUG 
config DEBUG_SYSTEM
endif

**Comment by pradeepns85** [comment]
@sunghan-chang , I have taken care of the comments and named the folder as  apps/system/sysdbg/
Kindly review the latest changes

**Comment by seinfra** [comment]
Code Rule Check OK.

File: os/Kconfig.debug [hunk 19]
```diff
@@ -529,6 +529,46 @@ config DEBUG_TTRACE
 	---help---
 		Enable T-trace debug SYSLOG output (disabled by default)
 
+comment "System Debug Options"
+
+config DEBUG_SYSTEM
+	bool "System debug"
+	default n
+	depends on DEBUG_VERBOSE
```
Comment (sunghan-chang) [inline os/Kconfig.debug]: 1. Why do you make a dependancy with DEBUG_VERBOSE? You use "dbg" and it is enabled when DEBUG_ERROR is enabled.
2. Let's use "select" instead of "depends on". "select" is better to use and DEBUG_VERBOSE doesn't have any dependancy so that we can use "select".

File: os/include/tinyara/debug/sysdbg.h [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+#define __INCLUDE_DEBUG_SYSDBG_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <tinyara/sched.h>
+#include <tinyara/clock.h>
+
```
Comment (sunghan-chang) [inline os/include/tinyara/debug/sysdbg.h]: Let's include below:
<sys/types.h> to use pid_t.
<stdint.h> to use uint8_t
<semaphore.h> to use sem_t

File: os/kernel/debug/sysdbg.c [hunk 21]
```diff
@@ -0,0 +1,933 @@
...
+ * Included Files
+ ****************************************************************************/
+
+#include <errno.h>
+#include <tinyara/config.h>
+#include <tinyara/arch.h>
+#include <tinyara/kmalloc.h>
+#include <tinyara/debug/sysdbg.h>
+#include "sched/sched.h"
+
```
Comment (sunghan-chang) [inline os/kernel/debug/sysdbg.c]: include <sys/types.h> to size_t
<stdint.h> to use uintxx_t
<tinyara/fs/fs.h> to struct file
<stdbool.h> to bool
<assert.h> to panic
<semaphore.h> to sem_t

File: apps/system/sysdbg/Makefile [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+#
+###########################################################################
+
+-include $(TOPDIR)/.config
+-include $(TOPDIR)/Make.defs
+include $(APPDIR)/Make.defs
+
+# System Debugging Application
+
+APPNAME = sysdbgapp
```
Comment (sunghan-chang) [inline apps/system/sysdbg/Makefile]: Let's remove

**Comment by pradeepns85** [comment]
@sunghan-chang , I have addressed the comments and also changed  some of the variables name to match the implementaion. Kindly review it.

File: os/Kconfig.debug (hunk 15 above)
Comment (pradeepns85) [reply os/Kconfig.debug]: Ok. Done. I have selected DEBUG_ERROR

File: os/include/tinyara/debug/sysdbg.h (hunk 20 above)
Comment (pradeepns85) [reply os/include/tinyara/debug/sysdbg.h]: Ok Done

**Comment by sunghan-chang** [comment]
@juitem @samirsahu1947 Could you review this?
